spec/lib/pwned_password_downloader_spec.rb

```diff
@@ -0,0 +1,34 @@
+require 'spec_helper'
+require 'pwned_password_downloader'
+require 'tempfile'
+
+RSpec.describe PwnedPasswordDownloader do
+  let(:destination) { Dir.mktmpdir('pwned_passwords') }
+  subject(:downloader) { PwnedPasswordDownloader.new(destination:) }
+
+  before do
+    stub_request(:get, URI.join(PwnedPasswordDownloader::RANGE_API_ROOT, '00000').to_s).to_return(
+      body: "0005AD76BD555C1D6D771DE417A4B87E4B4:10\r\n000A8DAE4228F821FB418F59826079BF368:4",
+    )
+    stub_request(:get, URI.join(PwnedPasswordDownloader::RANGE_API_ROOT, '00001').to_s).
+      to_return(body: '0005DE2A9668A41F6A508AFB6A6FC4A5610:1')
+    allow(Thread).to receive(:new).and_yield(rand.to_s)
+    allow(downloader).to receive(:queue).and_return([])
```

Comment:
I think we should let the code test w/ real threads & queues, I wouldn't bother stubbing these, or if we did, just set `num_threads` to something lower

Reply:
> I think we should let the code test w/ real threads & queues, I wouldn't bother stubbing these

If we did this, what would you suggest to do to know when the results are ready to assert against?

Reply:
its a synchronous method so it'll still wait on `queue.empty?` to finish before returning

Reply:
A problem I ran into is that the `queue.empty?` check will pass when one of the threads picks up the last queue, potentially before it's finished writing the file for that range. I suspect that could also happen as a race condition in real-world usage. We'd probably want to change the `queue.empty?` check to check something that assures the last file is written, maybe we can use `progress_bar.finished?` for that.

Reply:
ed5043803 improves the issue I described in my previous comment.

With these changes, the stubs for the threads can be removed, but I noticed it seems to increase the runtime of the specs by about 0.5s (~0.55s vs ~0.1s), regardless of `num_threads`.